src/dune_rules/alias_rec.ml

```diff
@@ -1,23 +1,122 @@
+end
+
+module Source_tree_only = struct
+  let dep_on_alias_if_exists alias =
+    let open Action_builder.O in
+    let* definition = Action_builder.of_memo (Load_rules.alias_exists alias) in
+    match definition with
+    | false -> Action_builder.return Alias_status.Not_defined
+    | true ->
+      let+ () = Action_builder.alias alias in
+      Alias_status.Defined
+
   module Map_reduce =
-    Source_tree.Dir.Make_map_reduce
-      (Action_builder)
-      (Action_builder.Alias_status)
+    Source_tree.Dir.Make_map_reduce (Action_builder) (Alias_status)
 
-  let traverse dir ~f =
+  let dep_on_alias_rec name dir =
     let ctx_name, src_dir = Path.Build.extract_build_context_exn dir in
     let build_dir = Context_name.build_dir (Context_name.of_string ctx_name) in
     let f dir =
       let path =
         Path.Build.append_source build_dir (Source_tree.Dir.path dir)
       in
-      f ~path
+      dep_on_alias_if_exists (Alias.make ~dir:path name)
     in
     let open Action_builder.O in
     Source_tree.find_dir src_dir |> Action_builder.of_memo >>= function
-    | None -> Action_builder.return Action_builder.Alias_status.Not_defined
+    | None -> Action_builder.return Alias_status.Not_defined
     | Some src_dir ->
       Map_reduce.map_reduce src_dir ~traverse:Sub_dirs.Status.Set.normal_only ~f
-end)
+end
+
+module Traverse_build_dirs = struct
+  module Lookup_alias = struct
+    type result =
+      { alias_exists : Alias_status.t
+      ; allowed_subdirs : Filename.Set.t
+      }
+
+    let of_dir_set ~exists dirs =
+      let allowed_subdirs =
+        match Dir_set.toplevel_subdirs dirs with
+        | Infinite -> Filename.Set.empty
+        | Finite sub_dirs -> sub_dirs
+      in
+      { alias_exists = exists; allowed_subdirs }
+  end
+
+  let dep_on_alias_if_exists alias =
+    let open Action_builder.O in
+    let* load_rules_result =
+      Action_builder.of_memo
+        (Load_rules.load_dir ~dir:(Path.build (Alias.dir alias)))
+    in
+    match load_rules_result with
+    | Source _ | External _ ->
+      Code_error.raise "Alias in a non-build dir"
+        [ ("alias", Alias.to_dyn alias) ]
+    | Build { aliases; allowed_subdirs; rules_here = _ } -> (
+      match Alias.Name.Map.find aliases (Alias.name alias) with
+      | None ->
+        Action_builder.return
+          (Lookup_alias.of_dir_set ~exists:Not_defined allowed_subdirs)
+      | Some _ ->
+        Action_builder.alias alias
+        >>> Action_builder.return
+              (Lookup_alias.of_dir_set ~exists:Defined allowed_subdirs))
+    | Build_under_directory_target _ ->
+      Action_builder.return
+        { Lookup_alias.alias_exists = Not_defined
+        ; allowed_subdirs = Filename.Set.empty
+        }
+
+  let dep_on_alias_rec name dir =
+    let rec map_reduce dir ~f =
+      let open Action_builder.O in
+      let* should_traverse =
+        match Path.Build.drop_build_context dir with
+        | None -> Action_builder.return true
```

Comment:
Shouldn't dune error out in this case? Unlike the `| None -> true` branch in line 104, in this case it has not been possible to remove the `_build` prefix. I fail to see in which cases this would be expected.

Reply:
This means it’s not in source, and we should traverse it